File: back/src/controller/ingredientesReceta.controller.js
const db = require('../db/conection');

const getIngredientesRecetas = async (req, res) => { 
  const usuarioId = req.session.id;
  console.log(`ID de usuario: ${usuarioId}`);
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  } 

  const sql = `SELECT * FROM ingredientes_por_receta WHERE usuario_id = ?`;
  try {
    const [ingredientes] = await db.query(sql, [usuarioId]);
    if (ingredientes.length === 0) {
      return res.status(404).json({ mensaje: 'No se encontraron ingredientes para este usuario' });
    }
    return res.status(200).json(ingredientes);
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const getUnIngredienteReceta = async (req, res) => {
  const { id } = req.params; // <-- Cambiado a 'id'
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }   
  if (!id) {
    return res.status(400).json({ error: 'Falta el ID del ingrediente' });
  }
  const sql = 'SELECT * FROM ingredientes_por_receta WHERE id = ? AND usuario_id = ?';
  try {
    const [ingrediente] = await db.query(sql, [id, usuarioId]);
    if (ingrediente.length === 0) {
      return res.status(404).json({ error: 'Ingrediente no encontrado o no tienes permiso para verlo' });
    }
    return res.status(200).json(ingrediente[0]);
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};    

const postIngredienteReceta = async (req, res) => {
  const { recetaId, ingredienteId, peso } = req.body;
  const usuarioId = req.session.id;
  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }

  if (!recetaId || !ingredienteId || !peso) { 
    return res.status(400).json({ error: 'Faltan datos obligatorios' });
  }
  const sql = 'INSERT INTO ingredientes_por_receta (receta_id, ingrediente_id, peso, usuario_id) VALUES (?, ?, ?, ?)';
  try {
    const [result] = await db.query(sql, [recetaId, ingredienteId, peso, usuarioId]);
    return res.status(201).json({ mensaje: 'Ingrediente agregado correctamente', id: result.insertId });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const putIngredienteReceta = async (req, res) => {
  const { id } = req.params; // <-- Cambiado a 'id'
  const { peso } = req.body;
  const usuarioId = req.session.id;

  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }

  if (!id || !peso) {
    return res.status(400).json({ error: 'Faltan datos obligatorios' });
  }

  const sql = 'UPDATE ingredientes_por_receta SET peso = ? WHERE id = ? AND usuario_id = ?';
  try {
    const [result] = await db.query(sql, [peso, id, usuarioId]);
    if (result.affectedRows === 0) {
      return res.status(404).json({ error: 'Ingrediente no encontrado o no tienes permiso para actualizarlo' });
    }
    return res.status(200).json({ mensaje: 'Ingrediente actualizado correctamente' });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

const deleteIngredienteReceta = async (req, res) => {
  const { id } = req.params; // <-- Cambiado a 'id'
  const usuarioId = req.session.id;

  if (!usuarioId) {
    return res.status(401).json({ error: 'No estás autenticado' });
  }

  if (!id) {
    return res.status(400).json({ error: 'Falta el ID del ingrediente' });
  }

  const sql = 'DELETE FROM ingredientes_por_receta WHERE id = ? AND usuario_id = ?';
  try {
    const [result] = await db.query(sql, [id, usuarioId]);
    if (result.affectedRows === 0) {
      return res.status(404).json({ error: 'Ingrediente no encontrado o no tienes permiso para eliminarlo' });
    }
    return res.status(200).json({ mensaje: 'Ingrediente eliminado correctamente' });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ error: 'Error interno del servidor' });
  }
};

module.exports = {
  getIngredientesRecetas,
  getUnIngredienteReceta,
  postIngredienteReceta,
  putIngredienteReceta,
  deleteIngredienteReceta
};

File: ListaDeLaCompra_copia/back_py/controller/ingredientes_receta_controller.py
from db.conexion import get_connection
from flask import session

def get_ingredientes_receta_por_usuario():
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    sql = "SELECT * FROM ingredientes_por_receta WHERE usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (usuario_id,))
        ingredientes_list = cursor.fetchall()
        if not ingredientes_list:
            return {'mensaje': 'No se encontraron ingredientes para este usuario'}, 404
        return ingredientes_list, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def get_ingredientes_por_receta(receta_id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not receta_id:
        return {'error': 'Falta el ID de la receta'}, 400
    sql = "SELECT * FROM ingredientes_por_receta WHERE receta_id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (receta_id, usuario_id))
        ingredientes_list = cursor.fetchall()
        if not ingredientes_list:
            return {'error': 'No se encontraron ingredientes para esta receta'}, 404
        return ingredientes_list, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def get_un_ingrediente_receta(id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not id:
        return {'error': 'Falta el ID del ingrediente'}, 400
    sql = "SELECT * FROM ingredientes_por_receta WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor(dictionary=True)
        cursor.execute(sql, (id, usuario_id))
        ingrediente_dict = cursor.fetchone()
        if not ingrediente_dict:
            return {'error': 'Ingrediente no encontrado o no tienes permiso para verlo'}, 404
        return ingrediente_dict, 200
    except Exception:
        return {'error': 'Error interno del servidor'}, 500

def post_ingrediente_receta(receta_id, ingrediente_id, peso):
    usuario_id = session.get('usuario_id')
    print
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not all([receta_id, ingrediente_id, peso]):
        return {'error': 'Faltan datos obligatorios'}, 400
    sql = "INSERT INTO ingredientes_por_receta (receta_id, ingrediente_id, peso, usuario_id) VALUES (%s, %s, %s, %s)"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (receta_id, ingrediente_id, peso, usuario_id))
        db.commit()
        return {'mensaje': 'Ingrediente agregado correctamente', 'id': cursor.lastrowid}, 201
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def post_varios_ingredientes_receta(ingredientes):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not isinstance(ingredientes, list) or len(ingredientes) == 0:
        return {'error': 'Faltan datos obligatorios'}, 400
    sql = "INSERT INTO ingredientes_por_receta (receta_id, ingrediente_id, peso, usuario_id) VALUES (%s, %s, %s, %s)"
    db = get_connection()
    cursor = db.cursor()
    try:
        db.start_transaction()
        for ingrediente in ingredientes:
            receta_id = ingrediente.get('recetaId')
            ingrediente_id = ingrediente.get('ingredienteId')
            peso = ingrediente.get('peso')
            cursor.execute(sql, (receta_id, ingrediente_id, peso, usuario_id))
        db.commit()
        return {'mensaje': 'Ingredientes agregados correctamente'}, 201
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def put_ingrediente_receta(id, peso):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not id or not peso:
        return {'error': 'Faltan datos obligatorios'}, 400
    sql = "UPDATE ingredientes_por_receta SET peso = %s WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (peso, id, usuario_id))
        db.commit()
        if cursor.rowcount == 0:
            return {'error': 'Ingrediente no encontrado o no tienes permiso para actualizarlo'}, 404
        return {'mensaje': 'Ingrediente actualizado correctamente'}, 200
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500

def delete_ingrediente_receta(id):
    usuario_id = session.get('usuario_id')
    if not usuario_id:
        return {'error': 'No estás autenticado'}, 401
    if not id:
        return {'error': 'Falta el ID del ingrediente'}, 400
    sql = "DELETE FROM ingredientes_por_receta WHERE id = %s AND usuario_id = %s"
    try:
        db = get_connection()
        cursor = db.cursor()
        cursor.execute(sql, (id, usuario_id))
        db.commit()
        if cursor.rowcount == 0:
            return {'error': 'Ingrediente no encontrado o no tienes permiso para eliminarlo'}, 404
        return {'mensaje': 'Ingrediente eliminado correctamente'}, 200
    except Exception:
        db.rollback()
        return {'error': 'Error interno del servidor'}, 500
